contracts/components/staking/SlashingController.sol

```diff
@@ -0,0 +1,438 @@
+            revert MissingRole(SLASHING_ARBITER_ROLE, msg.sender);
+        } else if (isInState(_proposalId, uint256(SlashStates.REVIEWED)) && !hasRole(SLASHER_ROLE, msg.sender)) {
+            revert MissingRole(SLASHER_ROLE, msg.sender);
+        }
+        // If it's in another state, _transitionTo() will revert
+    }
+
+    // Private param setting
+
+    function _setSlashingExecutor(ISlashingExecutor _executor) private {
+        if (address(_executor) == address(0)) revert ZeroAddress("_executor");
+        slashingExecutor = _executor;
+        emit SlashingExecutorChanged(address(_executor));
+    }
+
+    function _setStakingParametersManager(FortaStakingParameters _stakingParameters) private {
+        if (address(_stakingParameters) == address(0)) revert ZeroAddress("_stakingParameters");
+        stakingParameters = _stakingParameters;
+        emit StakingParametersManagerChanged(address(_stakingParameters));
+    }
+
+    function _setDepositAmount(uint256 _amount) private {
+        if (_amount == 0) revert ZeroAmount("_amount");
+        depositAmount = _amount;
+        emit DepositAmountChanged(depositAmount);
+    }
+
+    function _setSlashPercentToProposer(uint256 _amount) private onlyValidPercent(_amount) {
+        slashPercentToProposer = _amount;
+        emit SlashPercentToProposerChanged(_amount);
+    }
+
+    function _setSlashPenalties(bytes32[] calldata _slashReasons, SlashPenalty[] calldata _slashPenalties) private {
+        uint256 length = _slashReasons.length;
+        if (length != _slashPenalties.length) revert DifferentLenghtArray("_slashReasons", "_slashPenalties");
+        for (uint256 i = 0; i < length; i++) {
+            if (penalties[_slashReasons[i]].mode != PenaltyMode.UNDEFINED) {
+                emit SlashPenaltyRemoved(_slashReasons[i], penalties[_slashReasons[i]].percentSlashed, penalties[_slashReasons[i]].mode);
+            }
+            penalties[_slashReasons[i]] = _slashPenalties[i];
+            emit SlashPenaltyAdded(_slashReasons[i], _slashPenalties[i].percentSlashed, _slashPenalties[i].mode);
+        }
+    }
+
+    // Evidence handling
+    function _submitEvidence(uint256 _proposalId, uint256 _stateId, string[] calldata _evidence) private {
+        uint256 evidenceLength = _evidence.length;
+        if (evidenceLength == 0) revert ZeroAmount("evidence lenght");
+        if (evidenceLength > MAX_EVIDENCE_LENGTH) revert ArrayTooBig(evidenceLength, MAX_EVIDENCE_LENGTH);
+        emit EvidenceSubmitted(_proposalId, _stateId, _evidence);
+    }
+
+    // Private deposit handling
+    function _returnDeposit(uint256 _proposalId) private {
+        SafeERC20.safeTransfer(depositToken, proposals[_proposalId].proposer, deposits[_proposalId]);
+        deposits[_proposalId] = 0;
+        emit DepositReturned(_proposalId, proposals[_proposalId].proposer, deposits[_proposalId]);
```

Comment:
just in case the depositToken is ERC777, you may want to protect against reenttrancy by doing  
```suggestion
        uint256 amount = deposits[_proposalId];
        delete deposits[_proposalId];
        SafeERC20.safeTransfer(depositToken, proposals[_proposalId].proposer, amount);
        emit DepositReturned(_proposalId, proposals[_proposalId].proposer, deposits[_proposalId]);
```

Reply:
also, you emit the value after cleaning, I assume this was a mistake